apps/designer/app/canvas/features/text-editor/text-editor.tsx

```diff
@@ -39,6 +49,61 @@ const AutofocusPlugin = () => {
   return null;
 };
 
+const RemoveParagaphsPlugin = () => {
+  const [editor] = useLexicalComposerContext();
+
+  // register own commands before RichTextPlugin
+  // to stop propagation
+  useLayoutEffect(() => {
+    const removeCommand = editor.registerCommand(
+      KEY_ENTER_COMMAND,
+      (event) => {
+        const selection = $getSelection();
+        if (!$isRangeSelection(selection)) {
+          return false;
+        }
+        event?.preventDefault();
+        // returns true which stops propagation
+        return editor.dispatchCommand(INSERT_LINE_BREAK_COMMAND, false);
+      },
+      COMMAND_PRIORITY_EDITOR
+    );
+
+    // flatten pasted paragraphs
+    const removeNodeTransform = editor.registerNodeTransform(
+      RootNode,
+      (node) => {
+        // merge paragraphs into first with line breaks between
+        if (node.getChildrenSize() > 1) {
+          const children = node.getChildren();
+          let first;
+          for (let index = 0; index < children.length; index += 1) {
+            const paragraph = children[index];
+            if (paragraph instanceof ElementNode) {
+              if (index === 0) {
+                first = paragraph;
+              } else if (first) {
+                first.append($createLineBreakNode());
+                for (const child of paragraph.getChildren()) {
+                  first.append(child);
+                }
+                paragraph.remove();
+              }
+            }
+          }
+        }
+      }
+    );
```

Comment:
How does this work? How do we tell whether pasted content contains paragraphs? I would expect something like `isParagraph(node)` somewhere in this code.

Reply:
With default configuration root contains only paragraphs. Lexical converts headings to paragraphs on paste for example. So he we just check root children which are all paragraphs.

Reply:
Got it. Might be a good idea to add this as a code comment